### PR TITLE
Add proper backgroundjobs to OCP

### LINKED ISSUE #4818: BackgroundJob API is unusable
The public background job API is afaik used by specifying this in your info.xml

```xml
<background-jobs>
    <job>OCA\News\Cron\Updater</job>
</background-jobs>
```

Then the specified class has to implement this interface https://github.com/nextcloud/server/blob/master/lib/public/BackgroundJob/IJob.php

As you may have noticed the public interface is absolutely unusable, that's why there is https://github.com/nextcloud/server/blob/master/lib/private/BackgroundJob/Job.php

Anyone who wants to implement something using the cronjob API however has to duplicate your code because the class lives in the private namespace. Would it be possible to give us mortal app devs access to a saner interface which works with DI :D?

```php
interface IBackgroundJob {
    function run();
}
```

That can be used like this?

```php
class Updater implements OCP\BackgroundJob\IBackgroundJob {
    public function __construct(UpdateService $updater) {
        $this->updater = $updater;
    }

    public function run() {
        $this->updater->updateFeeds();
    }
}
```

@LukasReschke @MorrisJobke @ChristophWurst

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #4818

Finally a public API for backgroundjobs

Todo:
- [x] Some tests

### CODE REVIEW

File: lib/public/BackgroundJob/Job.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	/**
+	 * The actual function that is called to run the job
+	 *
+	 * @param $argument
+	 * @return mixed
+	 *
+	 * @since 15.0.0
+	 */
+	abstract protected function run($argument);
```
Comment (ChristophWurst) [inline lib/public/BackgroundJob/Job.php]: you'll get bonus points for something a `JobTester` that can be used like https://api.symfony.com/3.0/Symfony/Component/Console/Tester/CommandTester.html :wink:

File: lib/public/BackgroundJob/Job.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+/**
+ * Base class for background jobs
+ *
+ * This is here if you want to do advanced stuff in your background jobs.
+ * For the most common use cases have a look at QueuedJob and TimedJob
+ *
+ * @since 15.0.0
+ */
+abstract class Job implements IJob {
```
Comment (nickvergessen) [inline lib/public/BackgroundJob/Job.php]: should this extend the OC one, or the otherway around? Or is there any reason why they do not?

File: lib/public/BackgroundJob/Job.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/public/BackgroundJob/Job.php]: of course we can add that in a follow-up.

**Review by MorrisJobke — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]
Looks good! Great to have this in OCP :rocket:

**Pull request merged**